Enhance the new color picker design

jorgefilipecosta opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Part of https://github.com/WordPress/gutenberg/issues/34284.

Applies the design changes specified by @ciampo in order to try to match what was proposed in the mockups: 
  - Adjust the white border on the circle controls to match design (compare storybook with design mockup).
  - The input padding is off, not sure if this is something we need to affect on all inputs or we can just start with the input + slider case.
  - Spacing between slider and input is off.
  - Use the new settings icon from wp/icons.
  - the color format dropdown should be rendered without borders (not sure how we want to handle this at the component level since it should be an optional thing used rarely)
  - Adjust circular handle of slider since it's too big compared to designs

## Screenshots <!-- if applicable -->

| Before | After |
|---|---|
| <img width="226" alt="color-picker-before" src="https://user-images.githubusercontent.com/11271197/132250523-35a1940d-f60d-42f9-b35b-e962784adbec.png"> | <img width="234" alt="color-picker-after" src="https://user-images.githubusercontent.com/11271197/132250529-2504cb46-9836-4588-ad60-9ec8300776ab.png"> |

jorgefilipecosta committed:
Apply suggestions from code review

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com> (+1 squashed commit)
Squashed commits:
[eeb7fa834e] Implement new color picker design

jasmussen commented:
I guess #34287 needs to land before I can see this one outside of storybook. This is what I see:

![color](https://user-images.githubusercontent.com/1204802/132304528-1402cfbf-3d4b-4275-87e1-99f8bba44926.gif)

Nice, getting there. A few things:

- The white circle is set to 1.5px `border`, but the border can't render subpixels so it collapses to 1px. You can use box-shadow instead. Box shadows are removed in Windows high contrast mode, so be sure to combine it with a transparent `outline` or `border`, because transparent is made opaque in that mode.
- The inputs look good. But I wonder if we should be tweaking them here on a per-component basis, rather then fixing them at the source instead. Eventually all inputs should be the same, and the padding/margins should come from the input component itself. The focus style of the input also needs to be 1.5px rather than 2px, so lots of low hanging fruit there.

jasmussen commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -6,13 +6,37 @@ import styled from '@emotion/styled';
+
+export const InputWrapper = styled( InputControl )`
+	${ InputControlContainer } {
+		padding: 5px 8px;
```
I'm still wrapping my head around .ts files, but when I see bespoke paddings and margins like these it gets me a little worried about code drift. What's the best solution for keeping things the same here? I understand re-used SCSS variables (like the `$grid-unit`s we have) isn't possible here. Is it more of a matter of updating the source component used?

ciampo replied:
After switching to Emotion, we've been using the `space` util ([`packages/components/src/ui/utils/space.ts`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/ui/utils/space.ts)) in combination with some shared configuration ([`packages/components/src/utils/config-values.js`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/config-values.js)) — we should probably leverage them here as well, instead of these hardcoded values

(also cc'ing @sarayourfriend in case I missed anything)

jasmussen commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -6,13 +6,37 @@ import styled from '@emotion/styled';
+
+export const NumberControlWrapper = styled( NumberControl )`
+	${ InputControlContainer } {
+		width: 96px;
```
Same concern here about bespoke paddings.

On the width, what options do we have to use flex here so the items are automatically spaced and sized by the container?

ciampo replied:
The `NumberControlWrapper` is [rendered as a child of `<Spacer as={ HStack }>`](https://github.com/WordPress/gutenberg/blob/c3a5fe294b544e536617f34867b52a9e7e219be3/packages/components/src/ui/color-picker/input-with-slider.tsx#L28), which means that `Spacer` is effective a flex container, and that [props from `HStack`](https://github.com/WordPress/gutenberg/blob/e6e0ea8906c0fb01ea132e3386d5e42172614ba9/packages/components/src/h-stack/README.md) can be applied to that instance of `<Spacer />`. `NumberControlWrapper` can be treated as a flexbox child.

ciampo commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -6,13 +6,37 @@ import styled from '@emotion/styled';
 import InnerSelectControl from '../../select-control';
 import InnerRangeControl from '../../range-control';
 import { StyledField } from '../../base-control/styles/base-control-styles';
+import { ThumbWrapper } from '../../range-control/styles/range-control-styles'
```
```suggestion
import { ThumbWrapper } from '../../range-control/styles/range-control-styles';
```

ciampo commented on packages/components/src/ui/color-picker/input-with-slider.tsx:
```diff
@@ -27,7 +26,7 @@ export const InputWithSlider = ( {
 		<Spacer as={ HStack }>
-			<NumberControl
+			<NumberControlWrapper
 				__unstableInputWidth="5em"
```
I wonder if we can replace the `__unstableInputWidth` prop with some CSS directly from `styles.ts`

jorgefilipecosta committed:
Apply suggestions from code review

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com> (+1 squashed commit)
Squashed commits:
[eeb7fa834e] Implement new color picker design

jorgefilipecosta commented:
Hi @jasmussen, @ciampo,

Thank you a lot for the reviews, I tried to incorporate your feedback. Mainly by avoiding hardcoded values and reusing shared values and utils (that I was not aware of and really simplify things).
I also tried to make general changes and avoid things specific to this component.
For example, a general change I did with success was the slider and it seems to work for all sliders.

There is a missing general change that I tried and failed. It was to change the input height which should be 40 by default and currently is 30. But that general change did not work well and broke the design in other parts already in production:
![image](https://user-images.githubusercontent.com/11271197/132704698-45049848-a4f8-42da-ba89-45041387496e.png)


Although it seems we can easily change the height because we have constant for that in practice things don't work parts of the UI stop being centered etc (we should take this opportunity to improve our code and make sure in the future we can easily change the constant). Changing the input field height also probably requires changing things like button height otherwise things don't look nice side by side. So I guess that we should do that task in a different PR focused on it. Where we can properly review it and make sure we don't break any of the existing UI.

So, for now, I keep a style specifically for the color picker that sets the inputs to 40 so the color picker already uses the new design and after making the change general we can delete this one-line style.

Feel free to have a new look at this PR and let me know what other improvements we can do. Thank you in advance for any additional insights.

jasmussen commented:
It seems like input fields might be the next thing we handle, so we don't have to add too many edgecases like these. 👍 👍

ciampo commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -23,6 +40,18 @@ export const RangeControl = styled( InnerRangeControl )`
 	}
 `;
 
+// All inputs should be the same height so this should be changed at the component level.
+// That involves changing heights of multiple input types probably buttons too etc.
+// So until that is done we are already using the new height on the color picker so it matches the mockups.
```
Could you please open a new issue for this task, and add it to tracking issue #34284?

ciampo commented on packages/components/src/ui/color-picker/hex-input.tsx:
```diff
@@ -38,7 +38,12 @@ export const HexInput = ( { color, onChange, enableAlpha }: HexInputProps ) => {
 		<InputControl
 			__unstableInputWidth="8em"
```
I know this is not caused by this PR, but I wonder if we could refactor away from the `__unstableInputWidth` property in this case as well?

(cc @sarayourfriend in case there's a specific reason why we're using this prop)

sarayourfriend replied:
I don't have any insight into this prop, sorry!

ciampo replied:
That's fine! Thank you for replying anyway :)

jorgefilipecosta replied:
I removed the usage of __unstableInputWidth 👍

ciampo reviewed (commented):
When playing around in Storybook, I noticed how both "thumbs" (the circular dots used to pick a color in the "2d color square" and on the "saturation track") are not perfectly centre-aligned with their white border, and slightly wiggle when the other thumb is being dragged around. These issues are likely caused by sub-pixel rounding/rendering.

https://user-images.githubusercontent.com/1083581/133246420-d48ce755-f8df-4981-a71c-f5089f618e40.mp4




Apart from that, it'd be great to get @mtias and @pablohoneyhoney to take a look at this before merging.

mtias commented:
The sliders look mispositioned, not sure if it's an artefact of storybook alone:

![image](https://user-images.githubusercontent.com/548849/133266125-d55fd7ad-3adb-4f92-99e6-a1323e4eeffc.png)

The toggle button is a bit too large:

![image](https://user-images.githubusercontent.com/548849/133266379-9dec8d74-86a9-4710-b7d3-657d4522c339.png)

Most noticeable in toggled state:

| Current  | Design |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/548849/133266496-e4eb9417-bbb0-4742-bbbe-b18bdb5c0da2.png)  | ![image](https://user-images.githubusercontent.com/548849/133266537-c53cc049-a2d5-48ba-acb6-05d35d36c871.png)  |

The spacing between input and range control is also too tight:


| Current  | Design |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/548849/133266635-f565112f-6552-41bf-a288-2cd7a17ed5ac.png)  | ![image](https://user-images.githubusercontent.com/548849/133266697-e9869356-e0b4-4940-a694-18aca7414f7c.png)  |


There's something going on in the bottom edge of the picker where it resets values in rgb to 255. Not sure if it's just an issue with how the storybook is setup.

https://user-images.githubusercontent.com/548849/133267062-b5c19957-ee48-4d92-8c0c-1b947deb6ee6.mov

mtias commented:
Expanding on the video above, when the reset happens the opacity slider renders like this:

![image](https://user-images.githubusercontent.com/548849/133267983-a1e0748a-d926-4b4f-bb5b-3dac3fd0e9cd.png)

jorgefilipecosta commented:
> When playing around in Storybook, I noticed how both "thumbs" (the circular dots used to pick a color in the "2d color square" and on the "saturation track") are not perfectly centre-aligned with their white border, and slightly wiggle when the other thumb is being dragged around. These issues are likely caused by sub-pixel rounding/rendering.
> 
>  color-picker-sub-pixel.mp4 
> Apart from that, it'd be great to get @mtias and @pablohoneyhoney to take a look at this before merging.

Hi @ciampo,
I spend some time trying to debug this issue. Initially, I thought it may be caused by non-unitary borders and box shadows but even with unitary borders it still happens. This seems to be an issue with the chrome render.
I tested on firefox and safari and the issue does not happen there.
Firefox:
![image](https://user-images.githubusercontent.com/11271197/134062627-531782f9-8d2b-47af-80e6-7c213c13406f.png)


Safari:
![image](https://user-images.githubusercontent.com/11271197/134062580-e8841ba3-58ec-4507-a3bf-ad08dc002317.png)



The issue on chrome does not affects just our code even the react colorful samples are affected https://codesandbox.io/s/react-colorful-customization-demo-mq85z?file=/src/styles.css:
![image](https://user-images.githubusercontent.com/11271197/134061003-17866260-aa5b-4753-9ea1-17c4f8b1525b.png)

jorgefilipecosta commented:
Hi @mtias,

> The sliders look mispositioned, not sure if it's an artefact of storybook alone:


I made some updates and the issue seems fixed. Even on the storybook with big amplification things seem centered:
![image](https://user-images.githubusercontent.com/11271197/134070719-d9fff069-b2fa-4d88-96c2-ffab7000b09b.png)



> The toggle button is a bit too large:

The dimensions of the button were updated.

> The spacing between input and range control is also too tight:

Spacing was updated.

> There's something going on in the bottom edge of the picker where it resets values in rgb to 255. Not sure if it's just an issue with how the storybook is setup.

I was able to reproduce the issue, it does not seem to be an issue with this PR as it also happens on the trunk. The issue does not happen on react-colorful samples, it may be related to be a storybook setup. After merging this PR I will propose a PR to replace the current component with this one and we can check if the issue happens outside storybook. If it does we will need to debug it deeper and try to fix it.

mtias commented:
Excellent. Just a few lingering things:

- Let's use a 4px height for the range slider so the border radius can be rendered. (Also 9999px seems excessive?)
- There's a small jump when toggling the advanced controls.

youknowriad commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -53,9 +84,24 @@ export const ColorfulWrapper = styled.div`
+	${ inputHeightStyle }
+`;
+
+export const DetailsControlButton = styled( Button )`
```
It's a bit unfortunate that this is not a button that can be shown properly with just props. Why do we need to tweak instead of relying on composition of uncustomized Button (is the isSmall prop not enough?)

jorgefilipecosta replied:
The isSmall prop on buttons with just an icon creates a 36-pixel width per 24 pixels height. Here we want a 24x24 pixels button.

youknowriad replied:
> The isSmall prop on buttons with just an icon creates a 36-pixel width per 24 pixels height. Here we want a 24x24 pixels button.

Do you think we should fix it in the button's styles instead?

```
	&.is-small {
		height: 24px;
		line-height: 22px;
		padding: 0 8px;
		font-size: 11px;

		&.has-icon:not(.has-text) {
			padding: 0 8px;
			width: 24px;
		}
	}
```

According to this style from the button's styles, it's meant to be `24px` maybe there's bug?

jorgefilipecosta replied:
Good point, I will try to understand if it is a bug or not that the button appears with 36px, but the padding and other styles suggest it is intentional.

mtias commented:
A couple more, let's reconcile these two tooltips:

![image](https://user-images.githubusercontent.com/548849/134152403-2d773413-5117-40e9-84a8-4626a204328b.png)

![image](https://user-images.githubusercontent.com/548849/134152461-e8118ea2-8179-4e02-9a17-3f6890e669e8.png)

The border radius is correct on the first one, but we should use the darker version not the lighter gray.

youknowriad commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -53,9 +84,24 @@ export const ColorfulWrapper = styled.div`
+`;
+
+export const ColorHexInputControl = styled( InputControl )`
+	width: 8em;
```
I guess I have similar questions for all the customized lower level components: InputControl, SelectControl, NumberControl... All of these get specific styles here. While in some cases it's fine, I do wonder whether our components don't compose well enough.

In other words, if we struggle to use the base components ourselves to build higher-level ones, it's most likely that the same struggles will exist for folks trying to use the components in their apps (or other wordpress packages like block-editor, edit-post...)

youknowriad reviewed (approved):
This is looking good to me, I do worry a bit about all the specific customizations but this is not something new, just wondering if we could improve the situation.

ciampo commented:
> I spend some time trying to debug this issue. Initially, I thought it may be caused by non-unitary borders and box shadows but even with unitary borders it still happens. This seems to be an issue with the chrome render.

This is interesting — I experimented with [changing the `border-width` from `CONFIG.borderWidthFocus` (which is `1.5px`) to `2px` on the pointer](https://github.com/WordPress/gutenberg/blob/c2c78c4b8b395df4cc92743116edabb634a9a481/packages/components/src/ui/color-picker/styles.ts#L87) and that seemed to fix the issue for me, also in Chrome.

I wonder if we should change `CONFIG.borderWidthFocus` to be `2px`, also in light of [similar conversations](https://github.com/WordPress/gutenberg/pull/34712#discussion_r711052382) about sub-pixel rendering glitches

pablohoneyhoney commented:
Loving how this is progressing. A few details I'm seeing in Safari.

Here’s a shot re: shadows. 

<img width="194" alt="Screen Shot 2021-09-21 at 1 12 34 PM" src="https://user-images.githubusercontent.com/25516611/134208848-e2b3bfad-8613-4578-9042-ebe516d84f0a.png">

And how the handles move far from what was expected. Top is what I see in SB, and bottom what was in the Figma.

<img width="274" alt="Screen Shot 2021-09-21 at 5 34 23 PM" src="https://user-images.githubusercontent.com/25516611/134208615-65ec3595-3597-4f9c-8e3c-7507f02e1e08.png">

There’s a strange jump when clicking on the color handle.

![picker](https://user-images.githubusercontent.com/25516611/134208194-40e6ec78-9294-46b5-bd7d-21976fdc89f9.gif)

Fields in the Figma had #BBBBBB borders, and the slider had a #CCCCCC

What I see in SB.

<img width="318" alt="Screen Shot 2021-09-21 at 6 22 37 PM" src="https://user-images.githubusercontent.com/25516611/134209083-9cf3fd8f-71c2-411c-acbf-e54c0e93c0ac.png">

What was in Figma.

<img width="206" alt="Screen Shot 2021-09-21 at 6 22 21 PM" src="https://user-images.githubusercontent.com/25516611/134209041-01668e93-42e6-4ac6-a17f-3598e25718e6.png">

For some reason, the focus states and tooltips aren’t visible for me in Safari.

![picker2](https://user-images.githubusercontent.com/25516611/134209370-8ccf698a-f1bf-41bf-862c-c6d3a3db8201.gif)

The desired style for focus would like:

![image](https://user-images.githubusercontent.com/25516611/134210036-d29bb6d9-5eef-433a-9d6b-81748ca2dd4b.png)

ciampo commented on packages/components/src/ui/color-picker/styles.ts:
```diff
@@ -53,9 +91,26 @@ export const ColorfulWrapper = styled.div`
 	.react-colorful__pointer {
 		height: 16px;
 		width: 16px;
+		border: ${ CONFIG.borderWidthFocus } solid rgba( 255, 255, 255, 0 );
+		box-shadow: inset 0px 0px 0px ${ CONFIG.borderWidthFocus } #ffffff;
```
We should probably use the `box-shadow` technique for borders, as suggested by @jasmussen in https://github.com/WordPress/gutenberg/pull/34712#discussion_r713728714

jorgefilipecosta replied:
I'm not sure if this technique works for this case given that we need border and box-shadow at the same time.

jasmussen replied:
You can stack box shadows, that's what I did in https://github.com/WordPress/gutenberg/commit/2997a2f2b933b6d157862bf7ad50a67411abea79.

jorgefilipecosta commented:
> And how the handles move far from what was expected. Top is what I see in SB, and bottom what was in the Figma.
> 
> <img alt="Screen Shot 2021-09-21 at 5 34 23 PM" width="274" src="https://user-images.githubusercontent.com/25516611/134208615-65ec3595-3597-4f9c-8e3c-7507f02e1e08.png">


I made an update to make the circles not as far from the bar. Let me know if there are other improvements we should do.


> There’s a strange jump when clicking on the color handle.
> 
> ![picker](https://user-images.githubusercontent.com/25516611/134208194-40e6ec78-9294-46b5-bd7d-21976fdc89f9.gif)

That was a complex issue. I spend a considerable time debugging the component but it turns out the issue was not in the component. We had two pickers separated by a text displaying the color code in a container that made the three elements equally separated. It turns out moving the cursor changed the color code possibly changing the width of the text and that then repositioned the picker. I fixed the story to avoid these jumps.

> 
> Fields in the Figma had #BBBBBB borders, and the slider had a #CCCCCC
> 
> What I see in SB.
> 
> <img alt="Screen Shot 2021-09-21 at 6 22 37 PM" width="318" src="https://user-images.githubusercontent.com/25516611/134209083-9cf3fd8f-71c2-411c-acbf-e54c0e93c0ac.png">
> 
> What was in Figma.
> 
> <img alt="Screen Shot 2021-09-21 at 6 22 21 PM" width="206" src="https://user-images.githubusercontent.com/25516611/134209041-01668e93-42e6-4ac6-a17f-3598e25718e6.png">

Yes, the fields need to have a redesign that affects multiple fields and affects existing components. I think we should have an issue focused on the fields where we change its dimensions borders etc.


> For some reason, the focus states and tooltips aren’t visible for me in Safari.
> 
> ![picker2](https://user-images.githubusercontent.com/25516611/134209370-8ccf698a-f1bf-41bf-862c-c6d3a3db8201.gif)
> 
> The desired style for focus would like:
> 
> ![image](https://user-images.githubusercontent.com/25516611/134210036-d29bb6d9-5eef-433a-9d6b-81748ca2dd4b.png)

Yes, I was able to reproduce the issue in Safari. And the current focus style also does not matches the design. Let's try to have an issue focused on the range control where we address these issues given that the component is already in production.

jasmussen commented:
Will you follow up on the border to box-shadow? I'm happy to help here.

jorgefilipecosta commented:
Merging this PR so we can continue the work, the non addressed issues will be addressed in followups:
- Redesign the input fields taking into consideration the new border color and height (removing specific height from this component).
- Improve the RangeControl making the focus work on safari and implementing the new focus design.
- Address the glitch with circle borders/shadowns on chrome.
- Replace the current color component with this one.


If there is any follow-up task that I'm missing just let me know.

jorgefilipecosta commented:
@jasmussen any help there would be appreciated :)

jorgefilipecosta commented:
> > I spend some time trying to debug this issue. Initially, I thought it may be caused by non-unitary borders and box shadows but even with unitary borders it still happens. This seems to be an issue with the chrome render.
> 
> This is interesting — I experimented with [changing the `border-width` from `CONFIG.borderWidthFocus` (which is `1.5px`) to `2px` on the pointer](https://github.com/WordPress/gutenberg/blob/c2c78c4b8b395df4cc92743116edabb634a9a481/packages/components/src/ui/color-picker/styles.ts#L87) and that seemed to fix the issue for me, also in Chrome.

Hi @ciampo,
I did exactly the same experience and on my tests, the issue happens with 2px as the border.
![image](https://user-images.githubusercontent.com/11271197/134339843-b54f1efd-d6fc-4ab9-b382-e354b7ce0237.png)


Maybe it depends on the chrome version?

jasmussen commented:
Chrome rounds up subpixels to the nearest whole when applied to `border`, but not for box-shadow. There's a similar fun issue with border radii, where a 1.5px border radius renders as 1px: https://codepen.io/joen/pen/NWjLgJa?editors=1100

ciampo commented:
> Merging this PR so we can continue the work, the non addressed issues will be addressed in followups:
> 
> * Redesign the input fields taking into consideration the new border color and height (removing specific height from this component).
> * Improve the RangeControl making the focus work on safari and implementing the new focus design.
> * Address the glitch with circle borders/shadowns on chrome.
> * Replace the current color component with this one.
> 
> If there is any follow-up task that I'm missing just let me know.

@jorgefilipecosta I've update tracking issue https://github.com/WordPress/gutenberg/issues/34284 with the follow-up tasks you suggested. Let's use that issue to coordinate and update the status of the new color picker work
